### PR TITLE
Add a MLMHead layer

### PR DESCRIPTION
**Pull request by mattdangerw**
Fixes https://github.com/keras-team/keras-nlp/issues/129

### CODE REVIEW

File: keras_nlp/layers/mlm_classification_head.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+from tensorflow import keras
+
+
+class MLMClassificationHead(keras.layers.Layer):
+    """Masked Language Model classification head.
+
+    This layer takes two inputs:
+     - `inputs`: which should be a tensor of encoded tokens with shape
+            `(batch_size, sequence_length, encoding_dim)`.
+     - `mask_positions`: with be a tensor of integer positions to predict
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply keras_nlp/layers/mlm_classification_head.py:27]: Is `mask_positions` possibly being a RaggedTensor?

File: keras_nlp/layers/mlm_classification_head.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+    # Predict an output word for each masked input token.
+    outputs = keras_nlp.layers.MLMClassificationHead(
+        vocabulary_size=10000,
+        activation="softmax",
+    )(encoded_tokens, inputs["mask_indices"])
+
+    # Build a model
+    model = keras.Model(inputs, outpus)
+    model.compile(optimizer="adam", loss="sparse_categorical_crossentropy")
```
Comment (chenmoneygithub) [inline keras_nlp/layers/mlm_classification_head.py:81]: Can we show how a real data would be like? The actual training data has another field `mask_ids`, which is the label input to `model.fit`, representing the original tokens at positions getting masked out. Additionally, `mask_weights` would tell `model.fit()` mask padding information.

File: keras_nlp/layers/mlm_classification_head.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        x = self._dense(x)
+        x = self._layer_norm(x)
+
+        # Transform encodings to vocabulary_size predictions.
+        if self.embedding_weights is None:
+            outputs = tf.matmul(x, self._kernel)
+        else:
+            outputs = tf.matmul(
+                x,
+                tf.cast(self.embedding_weights, self.compute_dtype),
```
Comment (chenmoneygithub) [inline keras_nlp/layers/mlm_classification_head.py:173]: I am not sure about this - if here we do not cast x's dtype, would there be a type mismatch between `x` and `self.embedding_weights`?

**Review by chenmoneygithub — COMMENTED** [review]
Thanks! Mainly looks good, left some initial comments.

File: keras_nlp/layers/mlm_classification_head.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_nlp/layers/mlm_classification_head.py:27]: No ragged support yet, or if it's there it's incidental. We should look at that in a follow up issue.

File: keras_nlp/layers/mlm_classification_head.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    compute an MLM loss function.
+
+    Args:
+        vocabulary_size: The total size of the vocabulary for predictions.
+        embedding_weights: Optional. The weights of the word embedding used
+            to transform input token ids. The transpose of this weight matrix
+            will be used to project a token embedding vector to a prediction
+            over all input words, as described in [1].
+        inner_activation: The activation function of inner dense layer.
+        outer_activation: The activation function for the outputs of the layer.
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_nlp/layers/mlm_classification_head.py:48]: I was not sure, wanted to hear from @fchollet on this.

We could just do a boolean `return_logits`, but I wasn't sure if there was prior art for that. There's `from_logits` in metrics, but that is kind of different.

We should do whatever is the more keras-y thing here--boolean or flexible activation.

Are there legitimate use cases for softmax-like alternatives you might want to do here for activation?

File: keras_nlp/layers/mlm_classification_head.py (hunk 3 above)
Comment (mattdangerw) [reply keras_nlp/layers/mlm_classification_head.py:173]: In mixed precision, x will generally be float16 and embedding_weights float32, so omitting this leads to an error. Not sure if there is a better way to do this though.

Comment (chenmoneygithub) [reply keras_nlp/layers/mlm_classification_head.py:173]: Yea, I was wondering if we need to cast x's type as well? In the mixed precision case, compute_dtype would just be x's dtype, but would there be some case that x's dtype mismatch `self.compyte_dtype`?

Comment (mattdangerw) [reply keras_nlp/layers/mlm_classification_head.py:173]: Ah I believe the `autocast` functionality on the base Keras layer will handle that?

Comment (chenmoneygithub) [reply keras_nlp/layers/mlm_classification_head.py:173]: nice to know, thanks!

File: keras_nlp/layers/mlm_classification_head.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_nlp/layers/mlm_classification_head.py:48]: Maybe `intermediate_activation` and `activation`?

Then the core usage is similar to Dense().

```
MLMClassificationHead(
  vocabulary_size=100,
  activation="softmax",
)
```

**Comment by mattdangerw** [comment]
Let's rename this `MLMHead` so we can have a `ClassificationHead` in the future.

File: keras_nlp/layers/mlm_classification_head.py (hunk 2 above)
Comment (mattdangerw) [reply keras_nlp/layers/mlm_classification_head.py]: I switched this to real data and a manually computed loss, basically made this fully runnable in a colab.

I don't think we can show a full end to end training here, that is what our guides are for. We can link the guide from this layer when it is up.

File: keras_nlp/layers/mlm_classification_head.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_nlp/layers/mlm_classification_head.py]: I have updated with the change described above.

[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_nlp/layers/mlm_classification_head.py]: Confirmed with Francois, we'd like this approach.
